src/main/java/org/jsoup/select/NodeFilteringVisitor.java

```diff
@@ -0,0 +1,33 @@
+package org.jsoup.select;
+
+import org.jsoup.nodes.Node;
+
+/**
+ * Node visitor interface. Provide an implementing class to {@link NodeTraversor} to iterate through nodes.
+ * <p>
+ * This interface provides two methods, {@code head} and {@code tail}. The head method is called when the node is first
+ * seen, and the tail method when all of the node's children have been visited. As an example, head can be used to
+ * create a start tag for a node, and tail to create the end tag.
+ * </p>
+ */
+public interface NodeFilteringVisitor {
+    /**
+     * Callback for when a node is first visited.
+     *
+     * @param node the node being visited.
+     * @param depth the depth of the node, relative to the root node. E.g., the root node has depth 0, and a child node
+     * of that will have depth 1.
+     * @return {@code true} to continue visiting the subtree, {@code false} to remove it.
+     */
+    boolean head(Node node, int depth);
+
+    /**
+     * Callback for when a node is last visited, after all of its descendants have been visited.
+     *
+     * @param node the node being visited.
+     * @param depth the depth of the node, relative to the root node. E.g., the root node has depth 0, and a child node
+     * of that will have depth 1.
+     * @return {@code true} to keep the subtree, {@code false} to remove it.
+     */
+    boolean tail(Node node, int depth);
```

Comment:
I like this idea, and it should come at little overhead. It may then be feasible to really move this into `Node` and `Elements`.